controller/controller.go

```diff
@@ -0,0 +1,163 @@
+package controller
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/Informasjonsforvaltning/fdk-user-feedback-service/model"
+	"github.com/Informasjonsforvaltning/fdk-user-feedback-service/service"
+	"github.com/Informasjonsforvaltning/fdk-user-feedback-service/util"
+)
+
+type Controller interface {
+	CreateComment(w http.ResponseWriter, r *http.Request)
+	GetComments(w http.ResponseWriter, r *http.Request)
+	UpdateComment(w http.ResponseWriter, r *http.Request)
+	DeleteComment(w http.ResponseWriter, r *http.Request)
+}
+
+type ControllerImpl struct {
+	AuthService    service.AuthService
+	TopicIdService service.TopicIdService
+	TopicService   service.TopicService
+}
+
+func (controller *ControllerImpl) CreateComment(w http.ResponseWriter, r *http.Request) {
+	userId, statusCode := controller.AuthService.AuthenticateAndGetUserId(r.Header.Get("Authorization"))
+	if userId == nil || statusCode != http.StatusOK {
+		w.WriteHeader(statusCode)
+		return
+	}
+
+	entityId, statusCode := util.ParseRequestUrlPath(r.URL.Path)
+	if !util.SuccsessfulStatus(statusCode) || entityId == nil {
+		w.WriteHeader(statusCode)
+		return
+	}
+
+	post, err := util.DecodePost(r.Body)
+	if err != nil || post == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	topicId, err := controller.TopicIdService.GetTopicId(*entityId)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if topicId == nil {
+		createdTopic, statusCode := controller.TopicService.CreateTopic(*entityId)
+		if !util.SuccsessfulStatus(statusCode) || createdTopic == nil {
+			w.WriteHeader(statusCode)
+			return
+		}
+		topicId = &createdTopic.TopicId
+	}
+
+	created, statusCode := controller.TopicService.CreateTopicPost(model.Post{
+		PostId:   post.PostId,
+		UserId:   *userId,
+		TopicId:  *topicId,
+		Content:  post.Content,
+		ToPostId: post.ToPostId,
+	})
+	if !util.SuccsessfulStatus(statusCode) {
+		w.WriteHeader(statusCode)
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	json.NewEncoder(w).Encode(created)
+}
+
+func (controller *ControllerImpl) GetComments(w http.ResponseWriter, r *http.Request) {
+	entityId, statusCode := util.ParseRequestUrlPath(r.URL.Path)
+	if !util.SuccsessfulStatus(statusCode) || entityId == nil {
+		w.WriteHeader(statusCode)
+		return
+	}
+
+	topicId, err := controller.TopicIdService.GetTopicId(*entityId)
+	if err != nil || topicId == nil {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	topic, statusCode := controller.TopicService.GetTopic(*topicId)
```

Comment:
Kaller først GetTopicId og deretter GetTopic, kanskje det kan flyttes til en egen funksjon i TopicIdService som holder på denne logikken for å få en renere controller som kun kaller TopicIdService en gang?

Reply:
https://github.com/Informasjonsforvaltning/fdk-user-feedback-service/pull/1/commits/5e0566da56e181c374d7f6da93f87039cfc198cd